apps/fhir/server/authentication.py

```diff
@@ -8,30 +8,32 @@
 logger = logging.getLogger('hhs_server.%s' % __name__)
 
 
-def authenticate_crosswalk(crosswalk):
+def match_hicn_hash(hicn_hash):
     auth_state = FhirServerAuth(None)
     certs = (auth_state['cert_file'], auth_state['key_file'])
-
     # URL for patient ID.
     url = get_resourcerouter().fhir_url + \
         "Patient/?identifier=http%3A%2F%2Fbluebutton.cms.hhs.gov%2Fidentifier%23hicnHash%7C" + \
-        crosswalk.user_id_hash + \
+        hicn_hash + \
         "&_format=json"
     response = requests.get(url, cert=certs, verify=False)
     backend_data = response.json()
 
+    if backend_data.get('total', 0) > 1:
+        # Don't return a 404 because retrying later will not fix this.
+        raise UpstreamServerException("Duplicate beneficiaries found")
+
     if 'entry' in backend_data and backend_data['total'] == 1:
         fhir_id = backend_data['entry'][0]['resource']['id']
-        crosswalk.fhir_id = fhir_id
-        crosswalk.save()
+        return fhir_id, backend_data
 
-        logger.info("Success:Beneficiary connected to FHIR")
-        # Recheck perms
-    else:
-        if backend_data.get('total', 0) > 1:
-            # Don't return a 404 because retrying later will not fix this.
-            raise UpstreamServerException("Duplicate beneficiaries found")
+    raise exceptions.NotFound("The requested Beneficiary has no entry, however this may change")
+
+
+def authenticate_crosswalk(crosswalk):
+    fhir_id, backend_data = match_hicn_hash(crosswalk.user_id_hash)
```

Comment:
Do we still need this function? Can't this save be handled in the transaction of the create_beneficiary_record function?

Reply:
this is the function that covers the case where a `fhir_id` does not exist in the fhir server. I'd like to take a hard look at this distinct flow as a whole.

Reply:
Right. This case shouldn't happen. If it did, then BB2 should not allow the user to create a user record. I'm arguing to be strict about these cases.

Reply:
If this gets done the changes will either be added to https://github.com/CMSgov/bluebutton-web-server/pull/772 or live in their own PR.